File: opengllk/src/main/java/com/example/opengllk/refer/programs/HeightmapShaderProgram.java
/***
 * Excerpted from "OpenGL ES for Android",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material, 
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose. 
 * Visit http://www.pragmaticprogrammer.com/titles/kbogla for more book information.
***/
package com.example.opengllk.refer.programs;


import android.content.Context;
import android.opengl.GLES20;

import com.example.opengllk.R;

public class HeightmapShaderProgram extends ShaderProgram {
    private final int uMatrixLocation;
    private final int aPositionLocation;

    public HeightmapShaderProgram(Context context) {
        super(context, R.raw.heightmap_vertex_shader0,
            R.raw.heightmap_fragment_shader0);
        
        uMatrixLocation = GLES20.glGetUniformLocation(program, U_MATRIX);
        aPositionLocation = GLES20.glGetAttribLocation(program, A_POSITION);
    }

    public void setUniforms(float[] matrix) {
        GLES20.glUniformMatrix4fv(uMatrixLocation, 1, false, matrix, 0);
    }

    public int getPositionAttributeLocation() {
        return aPositionLocation;
    }
}

File: opengllk/src/main/java/com/example/opengllk/program/HeightMapShaderProgram.kt
package com.example.opengllk.program

import android.content.Context
import android.opengl.GLES20
import com.example.opengllk.R

/**
 * Created by Cry on 2018/2/6.
 */
class HeightMapShaderProgram(context: Context)
    : BaseShaderProgram(
        context,
        R.raw.heightmap_vertex_shader,
        R.raw.heightmap_fragment_shader
) {

    //定义需要链接的属性
    companion object {
        val A_POSITION = "a_Position"
        val U_MATRIX = "u_Matrix"
    }

    private val aPositionLocation = GLES20.glGetAttribLocation(programId, A_POSITION)

    private val uMatrixLocation = GLES20.glGetUniformLocation(programId, U_MATRIX)
    fun getPositionAttributeLocation() = aPositionLocation

    fun setUniforms(matrix: FloatArray) {
        GLES20.glUniformMatrix4fv(uMatrixLocation, 1, false, matrix, 0)
    }

}